File: 0002-add-two-numbers/0002-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        byte val1 = 0, val2 = 0, tmp, carrier = 0;
        ListNode curans = null;
        ListNode ans = new ListNode();
        boolean firsttime = true;
        boolean l1empty = false, l2empty = false;
        
        while (true)
        {

            // check if l1 has value
            if (l1 != null)
                val1 = (byte) l1.val;
            else
                l1empty = true;


            // check if l2 has value 
            if (l2 != null)
                val2 = (byte) l2.val;
            else
                l2empty = true;
            
            if (l1empty && l2empty && carrier != 1)
                break;

            tmp = (byte)(val1 + val2 + carrier);
            carrier = 0;
            if (tmp - (byte) 9 > 0)
            {
                carrier = 1;
                tmp = (byte) (tmp -  10);
            }
            
            // new a node and store value
            ListNode tmpnode = new ListNode(tmp);
            
            if (firsttime)
            {
                ans = tmpnode;
                curans = tmpnode;
                firsttime = false;
            }
            else
            {
                curans.next = tmpnode;
                curans = tmpnode;
            }

            // move to next l1, l2
            if (l1 != null)
                l1 = l1.next;
            if (l2 != null)
                l2 = l2.next;
            val1 = 0;
            val2 = 0;

        }

        return ans;
    }
}

File: 0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// 2 4 3
// 342
// 5 6 4
// 465

// 342+465 = 807

// 7 -> 0 -> 8

// l1 = [0], l2 = [0]
// 0

// l1 = [9,9,9,9,9,9,9], l2 = [9,9,9,9]
// [8,9,9,9,0,0,0,1]

// two pass: O(n)
// space: O(n)


// two pass: the first pass to find the longer one
// second pass: sum on the longer one, and attach one more node if needed 
// O(1) space

// [2,4,9]
// [5,6,4,9]

// l1Len
// 7 0 4 0 1

class Solution {
public:

    // each linked list is in the range [1, 100].
    int getLen(ListNode* l) {
        int result = 0;
        while (l) {
            result++;
            l = l->next;
        }
        return result;
    }

    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {

        int l1Len = getLen(l1);
        int l2Len = getLen(l2);

        ListNode* head = l1Len >= l2Len ? l1 : l2;
        ListNode* takePtr = head == l1 ? l2 : l1;
        int totalLen = head == l1 ? l1Len : l2Len;

        int carry = 0;
        ListNode* curr = head;
        while (totalLen) {
            int currVal = curr->val;
            if (takePtr) {
                currVal += takePtr->val;
                takePtr = takePtr->next;
            }
            currVal += carry;
            // cout << currVal << " " << l1Len << endl; 
            // 14 1
            // 1 4
            carry = currVal / 10;
            curr->val = currVal % 10;
            totalLen--;
            // deal with when we need to append a last node
            if (totalLen == 0) {
                if (carry) {
                    curr->next = new ListNode(carry);
                }
            }
            curr = curr->next;
        }

        return head;
    }
};